File: backend/src/main/java/com/stacktobasics/pokemoncatchbackend/domain/pokemon/Pokemon.java
package com.stacktobasics.pokemoncatchbackend.domain.pokemon;

import com.stacktobasics.pokemoncatchbackend.domain.AggregateRoot;
import lombok.*;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.annotation.Id;

import java.util.List;

import static com.stacktobasics.pokemoncatchbackend.domain.game.Game.UNUSED_GAMES;

@Slf4j
@Getter
@Setter
@NoArgsConstructor
@EqualsAndHashCode(of = "pokedexNumber")
public class Pokemon implements AggregateRoot {
    @Id
    private Integer pokedexNumber;
    private String name;
    private int generation;
    private EncounterDetails encounterDetails = new EncounterDetails();
    private int evolutionChainId;
    private boolean canBreed;
    private List<Integer> types;

    public Pokemon(@NonNull Integer pokedexNumber, @NonNull String name, int generation, boolean canBreed, @NonNull List<Integer> types) {
        this.pokedexNumber = pokedexNumber;
        this.name = name;
        this.generation = generation;
        this.canBreed = canBreed;
        this.types = types;
    }

    public void addEncounter(int catchRate, @NonNull String location, int gameId,
                             @NonNull String method, @NonNull List<String> conditions) {
         if(UNUSED_GAMES.contains(gameId)) return;
         encounterDetails.addEncounter(catchRate, location, gameId, method, conditions);
    }
}

File: frontend/types/Pokemon.ts
import EvolutionChain from "@/types/EvolutionChain";

export interface Location {
    name: string
    gameId: number
}

export interface Encounter {
    catchRate: number
    location: Location
    method: string
    conditions: string[]
}


export interface EncounterDetails {
    bestCatchRate: number
    encounters: Encounter[]
}

export default interface Pokemon {
    pokedexNumber: number
    name: string
    generation: number
    encounterDetails: EncounterDetails
    evolutionChainId: number
    types: number[]
}

export interface PokemonWithMeta extends Pokemon {
    owned: boolean
    catchable: boolean
    breedable: boolean
    catchAndBreed: boolean
}